libraries/standard/http/utest/http_send_utest.c

```diff
@@ -0,0 +1,1507 @@
+                                    NULL,
+                                    NULL,
+                                    0,
+                                    &response,
+                                    0 );
+
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, returnStatus );
+}
+
+/*-----------------------------------------------------------*/
+
+/* Test a null request headers buffer passed to the API. */
+void test_HTTPClient_Send_null_request_header_buffer( void )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    requestHeaders.pBuffer = NULL;
+    returnStatus = HTTPClient_Send( &transportInterface,
+                                    &requestHeaders,
+                                    NULL,
+                                    0,
+                                    &response,
+                                    0 );
+
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, returnStatus );
+}
+
+/*-----------------------------------------------------------*/
+
+/* Test a NULL response buffer passed to the API. */
+void test_HTTPClient_Send_null_response_buffer( void )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    response.pBuffer = NULL;
+    returnStatus = HTTPClient_Send( &transportInterface,
+                                    &requestHeaders,
+                                    NULL,
+                                    0,
+                                    &response,
+                                    0 );
+
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, returnStatus );
+}
+
+/*-----------------------------------------------------------*/
+
+/* Test the request headers not containing enough bytes for a valid status-line.
+ */
+void test_HTTPClient_Send_not_enough_request_headers( void )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    requestHeaders.headersLen = HTTP_MINIMUM_REQUEST_LINE_LENGTH - 1;
+    returnStatus = HTTPClient_Send( &transportInterface,
+                                    &requestHeaders,
+                                    NULL,
+                                    0,
+                                    &response,
+                                    0 );
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, returnStatus );
+}
+
+/*-----------------------------------------------------------*/
+
+/* Test a NULL request body but a non-zero requets body length.
+ */
+void test_HTTPClient_Send_null_request_body_nonzero_body_length( void )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    returnStatus = HTTPClient_Send( &transportInterface,
+                                    &requestHeaders,
+                                    NULL,
+                                    1,
+                                    &response,
+                                    0 );
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, returnStatus );
+}
+
+/*-----------------------------------------------------------*/
+
+/* Test when the Content-Length header cannot fit into the header buffer. */

```

Comment:
How is the Content-Length header not fitting into the header buffer in this test?

Reply:
The Content-Length is auto-written in HTTPClient_Send if 
`requestHeaders.bufferLen - requestHeaders.headersLen` can accommodate. 
I added some comments.